test_haystack/solr_tests/test_management_commands.py

```diff
@@ -183,6 +197,54 @@ def test_build_schema_wrong_backend(self):
         connections['whoosh']._index = self.ui
         self.assertRaises(ImproperlyConfigured, call_command, 'build_solr_schema', using='whoosh', interactive=False)
 
+    def test_build_schema(self):
+
+        # Stow.
+        oldhdf = constants.HAYSTACK_DOCUMENT_FIELD
+        oldui = connections['solr'].get_unified_index()
+        oldurl = settings.HAYSTACK_CONNECTIONS['solr']['URL']
+
+        needle = 'Th3S3cr3tK3y'
+        constants.HAYSTACK_DOCUMENT_FIELD = needle #Force index to use new key for document_fields
+        settings.HAYSTACK_CONNECTIONS['solr']['URL'] = settings.HAYSTACK_CONNECTIONS['solr']['URL'].rsplit('/',1)[0]+'/mgmnt'
+
+        ui = UnifiedIndex()
+        ui.build(indexes=[SolrMockSecretKeySearchIndex()])
+        connections['solr']._index = ui
+
+        renderedfile = StringIO()
+        orig_dir = get_script_dir()+"/confdir"
+        conf_dir = get_script_dir()+"/server/solr/server/solr/mgmnt/conf"
+        schemafile = conf_dir+'/schema.xml'
+        solrcfgfile = conf_dir+'/solrconfig.xml'
+
+        call_command('build_solr_schema',using='solr',stdout=renderedfile)
+        contents = renderedfile.getvalue()
+        self.assertTrue(contents.find("name=\"" + needle)!=-1)
+
+        call_command('build_solr_schema',using='solr',configure_dir=conf_dir)
+        with open(schemafile) as s:
+            self.assertTrue(s.read().find("name=\""+needle)!=-1)
+        with open(solrcfgfile) as s:
+            self.assertTrue(s.read().find("name=\"df\">"+needle)!=-1)
+        self.assertTrue(os.path.isfile(conf_dir+'/managed-schema.old'))
+
+        call_command('build_solr_schema',using='solr',reload=True)
+
+        os.rename(schemafile,schemafile+".bak")
+        while os.path.isfile(schemafile):#work around stupid delay in filesystem
```

Comment:
What filesystem was this running on that `os.rename` wasn't atomic? If this is necessary, I'd like to at the very least have a timeout so it can't run indefinitely